aiohttp/connector.py

```diff
@@ -560,28 +559,24 @@ def addrs(self):
 
     def add(self, host, addrs):
         self._addrs[host] = addrs
-        self._addrs_rr[host] = cycle(addrs)
 
         if self._ttl:
             self._timestamps[host] = monotonic()
 
     def remove(self, host):
         self._addrs.pop(host, None)
-        self._addrs_rr.pop(host, None)
 
         if self._ttl:
             self._timestamps.pop(host, None)
 
     def clear(self):
         self._addrs.clear()
-        self._addrs_rr.clear()
         self._timestamps.clear()
 
     def next_addrs(self, host):
-        # Return an iterator that will get at maximum as many addrs
-        # there are for the specific host starting from the last
-        # not itereated addr.
-        return islice(self._addrs_rr[host], len(self._addrs[host]))
+        addrs = self._addrs[host].copy()
+        shuffle(addrs)
+        return addrs
```

Comment:
Why not keep the original round robin but making sure that we get rid of the corner case ? such as

```
>>> addrs = cycle([1, 2, 3, 4])
>>> list(islice(addrs, 4))
[1, 2, 3, 4]
>>> next(addrs)
1
>>> list(islice(addrs, 4))
[2, 3, 4, 1]
```
Following your idea of using an ad-hoc list per call to `next_addrs` it's just a matter of make a next after built the list to make sure that we always use a shifted list per call.

Reply:
Because, anyway, cycle will start from actually random element. Why not to randomize explicitly ? Why we should stick on round-robin ?

```
In [7]: timeit list(islice(xxx, 4))
The slowest run took 15.44 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 447 ns per loop

In [8]: timeit shuffle(qwe)
The slowest run took 5.42 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 2.82 µs per loop
```

Well, it seems you are right....

Reply:
But no! if everyone will consume 4 items from slice, everytime everyone will receive exactly the same items, since every call to `nex_addrs` will rotate whole array.

Reply:
So, I fixed.